kamon-akka-http-server/src/main/scala/kamon/akka/http/server/instrumentation/ServerInstrumentation.scala

```diff
@@ -0,0 +1,158 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ========================================================== */
+package kamon.akka.http.server.instrumentation
+
+import akka.http.scaladsl.Http.{ IncomingConnection, ServerBinding }
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{ HttpResponse, HttpRequest }
+import akka.http.scaladsl.server.RequestContext
+import kamon.Kamon
+import kamon.akka.http.server.{ AkkaHttpServerExtension, AkkaHttpServer, AkkaHttpServerMetrics }
+import kamon.metric.Entity
+import kamon.trace.{ TraceContextAware, Tracer }
+import org.aspectj.lang.ProceedingJoinPoint
+import org.aspectj.lang.annotation._
+
+@Aspect
+class HttpRequestInstrumentation extends ServerInstrumentationUtils {
+
+  @DeclareMixin("akka.http.scaladsl.model.HttpRequest")
+  def mixinTraceContextAwareToRequest: TraceContextAware = TraceContextAware.default
+
+  @Pointcut("execution(akka.http.scaladsl.model.HttpRequest.new(..)) && this(request)")
+  def httpRequestCreation(request: HttpRequest with TraceContextAware): Unit = {}
+
+  @After("httpRequestCreation(request)")
+  def afterHttpRequestCreation(request: HttpRequest with TraceContextAware): Unit = {
+    val akkaHttpServerExtension = Kamon(AkkaHttpServer)
+    val incomingContext = Tracer.currentContext
+
+    if (incomingContext.isEmpty) {
+      val defaultTraceName = akkaHttpServerExtension.generateTraceName(request)
+      val token = request.headers.find(_.name == akkaHttpServerExtension.settings.traceTokenHeaderName).map(_.value)
+
+      val newContext = Kamon.tracer.newContext(defaultTraceName, token)
+      Tracer.setCurrentContext(newContext)
+    }
+
+    request.traceContext
+
+    recordAkkaHttpServerMetrics(request, akkaHttpServerExtension)
+  }
+}
+
+@Aspect
+class RequestContextInstrumentation extends ServerInstrumentationUtils {
+
+  @DeclareMixin("akka.http.scaladsl.server.RequestContextImpl")
+  def mixinTraceContextAwareToRequestContext: TraceContextAware = TraceContextAware.default
+
+  @Pointcut("execution(akka.http.scaladsl.server.RequestContextImpl.new(..)) && this(context)")
+  def requestContextCreation(context: RequestContext with TraceContextAware): Unit = {}
+
+  @Around("requestContextCreation(context)")
+  def afterCreation(pjp: ProceedingJoinPoint, context: RequestContext with TraceContextAware): Unit = {
+    val incomingContext = Tracer.currentContext
+
+    Tracer.withContext(incomingContext) {
+      pjp.proceed()
+    }
+  }
```

Comment:
This little part here is not necessary at all. If you run it, it is basically taking the current context, replacing it with itself, running the `pjp.proceed()` and then set the current context to what it was before, which is the same context, again.